File: Day31/Part4/bitmask_quest.cpp
#include <bits/stdc++.h>
using namespace std;

long long max_reward(vector<int> reward, vector<vector<int>> require){
    int n = reward.size();
    vector<int> need(n,0);
    for(int i=0;i<n;++i){
        int m=0; for(int p: require[i]) m|=1<<p; need[i]=m;
    }
    int N=1<<n; const long long NEG = -(1LL<<60);
    vector<long long> dp(N, NEG); dp[0]=0; long long ans=0;
    for(int mask=0; mask<N; ++mask){
        if(dp[mask]==NEG) continue; ans=max(ans, dp[mask]);
        for(int i=0;i<n;++i){
            if(mask>>i & 1) continue;
            if( (mask & need[i]) == need[i] ){
                int nmask = mask | (1<<i);
                dp[nmask] = max(dp[nmask], dp[mask] + reward[i]);
            }
        }
    }
    return ans;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<int> reward={3,5,2,10};
    vector<vector<int>> require={{},{0},{0},{1,2}};
    cout<<max_reward(reward, require)<<"\n"; // Expected 20
    return 0;
}

File: Day31/Part4/bitmask_quest.py
from typing import List

def max_reward(reward: List[int], require: List[List[int]]) -> int:
    n = len(reward)
    need_mask = [0]*n
    for i in range(n):
        m = 0
        for pre in require[i]:
            m |= 1<<pre
        need_mask[i] = m
    N = 1<<n
    dp = [-10**15]*N
    dp[0] = 0
    ans = 0
    for mask in range(N):
        if dp[mask] < -10**14:
            continue
        ans = max(ans, dp[mask])
        for i in range(n):
            if (mask>>i) & 1:
                continue
            if (mask & need_mask[i]) == need_mask[i]:
                nmask = mask | (1<<i)
                dp[nmask] = max(dp[nmask], dp[mask] + reward[i])
    return ans

if __name__ == "__main__":
    reward = [3,5,2,10]
    require = [[], [0], [0], [1,2]]
    print(max_reward(reward, require))  # Expected 20
